### PR TITLE
Drop `php-cs-fixer/php-cs-fixer` in favor of `nikic/php-parser`

### PR DESCRIPTION
**Pull request by ghostwriter**
Hi,

I would like to apologize for the recent experience you had to go through due to the act of sabotage and manipulation towards this project, caused by such unethical actions.

This patch combats https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7343, via the following changes:

- Replace PHP CS Fixer with PHP-Parser
- RemoveFinalKeyword Implementation
- RemoveReadOnlyKeyword Implementation
- Add `@final` DocBlock via `--mark-final` 
- Add `@readonly` DocBlock via `--mark-readonly`

It's important to note that these actions were deemed necessary due to the unethical nature and (in my opinion oppressive behavior) their violation of the spirit of open-source.

I've enabled `Allow edits by maintainers`; feel free to review the changes, provide any feedback, or make changes directly.

Thank you for your attention.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -15,7 +15,8 @@
         }
     ],
     "require": {
-        "php": "^8.1"
+        "php": "^8.1",
+        "nikic/php-parser": "^4.17"
```
Comment (ghostwriter) [inline composer.json:19]: `"friendsofphp/php-cs-fixer"` needs to be removed from `"require-dev"` once the test have been updated.

**Comment by mason-rogers** [comment]
I think you forgot something to add to the composer.json 😉 

```php
"conflict": {
    "friendsofphp/php-cs-fixer": "*"
},
```

You go low, I go lower.

**Comment by mason-rogers** [comment]
Also, readme contains two references to PHP CS Fixer

**Comment by stevebauman** [comment]
Thanks @ghostwriter, I really appreciate that you brought this PR up to me, but also re-wrote the implementation with a fix! Not a small feat ❤️ 

Luckily though, there was a PR that reverted the original change, so this ultimately won't be necessary: 

https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7348

Thanks again for the time you've spent on this. If the original change to PHP CS Fixer would have stayed, or is reversed again, I will merge this PR 👍

**Comment by Wirone** [comment]
@stevebauman personally I encourage you to merge this PR anyway. It drops Fixer as a dependency which in effect:
- solves the issue with it being used "against our will": we don't like the purpose of this package and would prefer to not involve Fixer in it, but we can only rely on your good will (or `conflict` 😆)
- produces smaller executable file, since Fixer has its own dependencies, while `nikic/php-parser` requires only few PHP extensions

I believe it's a great solution for both sides - you have smaller releases, almost independent code, while we don't grind our teeth 😉.

Of course it needs review and some tweaks, because `nikic/php-parser` should go to `require-dev` while `require` section should contain all extensions required by that lib (to work properly after PHAR is built).

**Comment by stevebauman** [comment]
> solves the issue with it being used "against our will": we don't like the purpose of this package and would prefer to not involve Fixer in it

Yes you've made your vendetta clear against this package.

>  but we can only rely on your good will (or conflict 😆)

Only because your PR was reverted.

> I believe it's a great solution for both sides - you have smaller releases, almost independent code, while we don't grind our teeth

No. This is what _you_ want, not me. I will not rewrite my project because you personally disagree with it ideologically.

You have to understand that there are those who will disagree with you and use PHP CS Fixer in ways you don't intend, or want them to. To attempt to police that is truly bizarre and unhealthy, to say the least. Let it go.

**Comment by Wirone** [comment]
I provided valid arguments in favour of this PR (not only ideological), but you decided to focus your answer on nonsense made out of thin air - OK, your choice. Goodnight!

**Pull request closed without merging**